drivers/rtc/rtc_rpi_pico.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (c) 2023 Andrew Featherstobne
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/irq.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/spinlock.h>
+
+#include <hardware/irq.h>
+#include <hardware/rtc.h>
+#include <hardware/regs/rtc.h>
+
+#define DT_DRV_COMPAT raspberrypi_pico_rtc
+
+/* struct tm start time:   1st, Jan, 1900 */
+#define TM_YEAR_REF 1900
+static int rtc_rpi_pico_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				       struct rtc_time *timeptr);
+struct rtc_rpi_pico_data {
+	struct k_spinlock lock;
+
+#ifdef CONFIG_RTC_ALARM
+	struct rtc_time alarm_time;
+	uint16_t alarm_mask;
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+	bool alarm_pending;
+#endif /* CONFIG_RTC_ALARM */
+};
+
+static struct rtc_rpi_pico_data rtc_data;
+
+LOG_MODULE_REGISTER(rtc_rpi, CONFIG_RTC_LOG_LEVEL);
+
+#ifdef CONFIG_RTC_ALARM
+static void rtc_rpi_isr(const struct device *dev)
+{
+	struct rtc_rpi_pico_data *data = dev->data;
+
+	rtc_alarm_callback callback;
+	void *user_data;
+
+	rtc_disable_alarm();
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	callback = data->alarm_callback;
+	user_data = data->alarm_user_data;
+	k_spin_unlock(&data->lock, key);
+
+	if (callback != NULL) {
+		callback(dev, 0, user_data);
+	} else {
+		data->alarm_pending = true;
+	}
+	/* re-enable the alarm. */
+	rtc_enable_alarm();
+}
+#endif
+
+static int rtc_rpi_pico_init(const struct device *dev)
```

Comment:
Shouldn't the clock for the RTC be enabled here somewhere?

Reply:
I created this PR before the clock controller work had been completed and merged onto `main`. Now that this is in place, I've rebased my work on the current `HEAD` of `main` and believe I have called the relevant parts of the `clock_control_` API.